grpc-api/src/main/java/io/deephaven/grpc_api/arrow/FlightServiceGrpcImpl.java

```diff
@@ -460,13 +460,12 @@ private void sealAndExport() {
                 }
 
                 // no more changes allowed; this is officially static content
-                resultTable.sealTable();
-                resultExportBuilder.submit(() -> {
+                resultTable.sealTable(() -> resultExportBuilder.submit(() -> {
                     // transfer ownership to submit's liveness scope, drop our extra reference
                     resultTable.manageWithCurrentScope();
                     resultTable.dropReference();
                     return resultTable;
-                });
+                }));
 
                 observer.onCompleted();
                 onRequestDone();
```

Comment:
I'm not familiar w/ the full code-paths; but typically, I see the onCompleted() calls being done inside the submit runnable. Is it correct to call onCompleted() outside of it here? (Also, I assume an onError is attached to the resultExportBuilder).

Reply:
This specific example is a little confusing because there are several different things going on at once.

I'm going to start with the simple cases you're talking about:
```
session.newExport(id)
    .require(dependency)
    .submit(() -> {
        observer.onNext(dependency.get().toProtobufType());
        observer.onComplete();
    });
```

You can't call onComplete until after having sent all data that you will send to that observer. Most of the time the data you want to send isn't available until after it has been resolved via the session's dependency mechanisms.

In the above example, that observer is supposed to ping a message every time it processes a Schema or RecordBatch from the user -- that way they know approximately how much is uploaded and whether or not it matches their expectations. Once we've processed the last RecordBatch we do not need that observer anymore.

However, I see your valid point that an error in resultExportBuilder won't get propagated to the user through that observer. They will find out about the failure through other mechanisms (dependency propagation, the export listener stream). But you're right, we should complete inside of the submit callback -- and the user will know when it's available if we do it that way.